File: gtci/12-tree-depth-first-search/cpp/1-minimum-path-sum.cpp
#include <algorithm>
#include <cassert>
#include <climits>
#include <iostream>
#include <limits>

#include "tree_utils.h"

class Solution {
    public:
    int getMinimumPathSum(TreeNode* root) {
        if (root == nullptr)
            return std::numeric_limits<int>::max();
        
        if (root->left == nullptr && root->right == nullptr)
            return root->val;

        return root->val + std::min(getMinimumPathSum(root->left), getMinimumPathSum(root->right));
    }
};

int main() {
    Solution sol;

    TreeNode* root = new TreeNode(10);
    root->left = new TreeNode(5);
    root->right = new TreeNode(5);
    root->left->left = new TreeNode(2);
    root->right->right = new TreeNode(1);
    assert(sol.getMinimumPathSum(root) == 16);

    delete root;

    std::cout << "All test cases passed.\n";

    return 0;
}

File: gtci/12-tree-depth-first-search/python3/1-minimum-path-sum.py
class TreeNode:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def get_minimum_path_sum(self, root):
        if root is None:
            return float("inf")

        if root.left is None and root.right is None:
            return root.val

        return root.val + min(
            self.get_minimum_path_sum(root.left), self.get_minimum_path_sum(root.right)
        )


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(10)
    root.left = TreeNode(5)
    root.right = TreeNode(5)
    root.left.right = TreeNode(2)
    root.right.right = TreeNode(1)
    assert sol.get_minimum_path_sum(root) == 16

    print("All test cases passed.")
